auth/auth0/src/auth0Authentication.ts

```diff
@@ -0,0 +1,127 @@
+import { fromEvent, FunctionEvent } from 'graphcool-lib'
+import { GraphQLClient } from 'graphql-request'
+import { decode } from 'jsonwebtoken'
+import * as fetch from 'isomorphic-fetch'
+import * as jwt from 'express-jwt'
+import * as jwksRsa from 'jwks-rsa'
+import * as pify from 'pify'
+
+interface User {
+  id: string
+}
+
+// https://auth0.com/docs/user-profile/normalized/auth0
+// with scope="openid email profile"
+interface Auth0User {
+  sub: string
+  email?: string
+  email_verified?: boolean
+  given_name?: string
+  family_name?: string
+  nickname?: string
+  name?: string
+  picture?: string
+}
+
+interface EventData {
+  accessToken: string
+}
+
+export default async (event: FunctionEvent<EventData>) => {
+  if (!process.env.AUTH0_DOMAIN || !process.env.AUTH0_AUDIENCE) {
+    console.log('Please provide a valid auth0 domain and audience!')
+    return { error: 'Auth0 Authentification not configured correctly.' }
+  }
+
+  try {
+    const graphcool = fromEvent(event)
+    const api = graphcool.api('simple/v1')
+    const { accessToken } = event.data
+
+    // check auth0 access token
+    await checkJwt(accessToken)
+
+    // call auth0 API to obtain user data
+    const auth0User = await getAuth0User(accessToken)
+    const auth0UserId = auth0User.sub
+
+    // get graphcool user with auth0UserId
+    const user: User = await getGraphcoolUser(api, auth0UserId)
+
+    // check if graphcool user exists, and create new one if not
+    const userId = user ? user.id : await createGraphcoolUser(api, auth0User)
+
+    // generate node token for User node
+    const token = await graphcool.generateNodeToken(userId!, 'User')
+    return { data: { id: userId, token } }
+  } catch (e) {
+    console.log(e)
+    return { error: 'An unexpected error occured during authentication.' }
+  }
+}
+
+async function checkJwt(accessToken: string) {
+  // Inject accessToken in a header to permit express middleware usage.
+  const req: any = { headers: { authorization: `Bearer ${accessToken}` } }
+
+  return pify(
```

Comment:
I wouldn't use an external lib for this. Just standard `return new Promise((resolve, reject) => { ... })`

Reply:
I started with a standard `new Promise`, it's just some sugar to make this prettier but if you don't like I can revert this pify implementation ;)

Reply:
I think it's better for more beginner developers to see what's going on without it.